src/protocol/jdbc/org/apache/jmeter/protocol/jdbc/AbstractJDBCTestElement.java

```diff
@@ -544,34 +552,15 @@ private String getStringFromResultSet(ResultSet rs) throws SQLException, Unsuppo
             results = new ArrayList<>();
             jmvars.putObject(currentResultVariable, results);
         }
-        int j = 0;
-        while (rs.next()) {
-            Map<String, Object> row = null;
-            j++;
-            for (int i = 1; i <= numColumns; i++) {
-                Object o = rs.getObject(i);
-                if(results != null) {
-                    if(row == null) {
-                        row = new HashMap<>(numColumns);
-                        results.add(row);
-                    }
-                    row.put(meta.getColumnLabel(i), o);
-                }
-                if (o instanceof byte[]) {
-                    o = new String((byte[]) o, ENCODING);
-                }
-                sb.append(o);
-                if (i==numColumns){
-                    sb.append('\n');
-                } else {
-                    sb.append('\t');
-                }
-                if (i <= varNames.length) { // i starts at 1
-                    String name = varNames[i - 1].trim();
-                    if (name.length()>0){ // Save the value in the variable if present
-                        jmvars.put(name+UNDERSCORE+j, o == null ? null : o.toString());
-                    }
-                }
+        int currentIterationIndex = 0;
+        int resultSetMaxRows = getIntegerResultSetMaxRows();
+        if (resultSetMaxRows < 0) {
+            while (rs.next()) {
+                currentIterationIndex = processRow(rs, meta, sb, numColumns, jmvars, varNames, results, currentIterationIndex);
+            }
+        } else {
+            while (currentIterationIndex < resultSetMaxRows && rs.next()) {
+                currentIterationIndex = processRow(rs, meta, sb, numColumns, jmvars, varNames, results, currentIterationIndex);
```

Comment:
yes, this is the first (initial) modification that covers all use cases with hsqldb and my use case (e.g. 99%).

Reply:
this is covered in JDBC_TESTS.jmx